framework agreement 3.0

lepistone opened the pull request:
Merge at the same time:
- [ ] OCA/vertical-ngo#151
- [ ] OCA/department#42
- [ ] OCA/vertical-ngo#187

lepistone committed:
pricelist selection filtered by portfolio

lepistone committed:
propagate data from pricelist to PO

lepistone committed:
new agreement pricelist behaviour

Includes a failed attempt to avoid rewriting the whole
_price_rule_get_multi. Let's be ugly for the moment, and explore a bit
more.

lepistone committed:
WIP

lepistone committed:
WIP screenshare

lepistone committed:
always show search_more (needs web module, TODO)

lepistone committed:
template fields on pricelist view

lepistone committed:
WIP available quantity

lepistone committed:
check agreed quantity

lepistone committed:
WIP purchase_exceptions

lepistone committed:
agreement exceptions now work

lepistone committed:
add related fields to prices

lepistone committed:
button to create a new pricelist

This automatically creates:
- a pricelist (i.e. an agreement)
- a pricelist version
- a pricelist rule
- a supplierinfo
- a partnerinfo for every product line in the agreement

lepistone committed:
restore PO states to the correct value

lepistone committed:
restore uniqueness and date order constraints

lepistone committed:
bring back tests on dates (but 1 skipped)

lepistone committed:
remove test_price_list because it is now irrelevant

Now framework agreements are real pricelists, so the functionality
tested here has been removed.

lepistone committed:
remove another test covered by standard pricelists

lepistone committed:
skip a test, and delete a duplicated one

The skipped one checks that we cannot change a price coming from an
agreement. Not sure we want to keep it.

lepistone committed:
skip tests on consumed quantity

I think we want to port those, but the data structure has changed.

lepistone committed:
new method to get the agreement line for a product

lepistone committed:
add missing import _

lepistone committed:
demove dead code for agreement class

It is a pricelist now.

lepistone committed:
remove stale print statement

lepistone committed:
test purchase_exceptions with framework_agreement

We had similar problems with sale_workflow, where workflow checks from
other modules are there, but the methods themselves are not accessible.
This looks like a bug in the registry, or the test code.

lepistone committed:
document new options in view

lepistone committed:
show new fields on purchase pricelists only

lepistone committed:
button to open prices of a pricelist

lepistone committed:
remove disabled test on onchange price

We decided to remove login from on_change methods where possible. We
might add back a check with purchase_exceptions in the future.

lepistone committed:
make portfolio start date required, enable test

lepistone committed:
enable tests on consumed quantity

lepistone committed:
adapt to non-required end date and more

lepistone committed:
portfolio name is a sequence now

lepistone committed:
help text end end date to say it's optional

lepistone committed:
add many related fields on prices

lepistone committed:
make the prices tree view editable

And take note of how we might be able to edit that stuff when grouped,
too.

lepistone committed:
since we have a line, the portfolio is running

lepistone committed:
add default for portfolio start date

lepistone committed:
show addresses on agreement, with onchange

The onchange methods are the same we have in the purchase order, see
multi_address modules in stock_logistic_transport.

lepistone committed:
if no portfolio, do not block on insufficient qty

lepistone committed:
make picking_type in FA view like in the PO

lepistone committed:
add debugging tip for weird onchange behaviour

lepistone committed:
propagate only non-empty fields agreement->po

This is done using getattr and setattr because read and write do not
work inside an onchange method.

lepistone committed:
use dictionary access onchange, thanks @yvaucher

lepistone committed:
hide partnerinfo related in the pricelist

This is because it gives a lot of trouble:

- by design and rather surprisingly, o2m-o2m related fields take only
  one record at each step, so they do not work as expected (they only
  follow the first element of the first o2m field).
- creating new partnerinfo seems to create duplicates.
- we cannot create lines here anyway.

lepistone committed:
allow to create new lines from 'open prices'

gurneyalex committed:
please flake 8 

by adding a blank line

gurneyalex commented:
@lepistone this is a huge diff... I think you should update framework_agreement/README.rst to explain what is in the new version. 

I'll wait until the WIP label is removed for a more in depth review.

lepistone committed:
remove duplicated field

lepistone committed:
make origin address required with a default

This is only for pricelists that are agreements. For that reason the
field is required in the view only, and it is filled up when
automatically generating a pricelist.

lepistone committed:
new agreement view, now more readable

Beware: before that commit the view was inherited, and it is not the
case anymore.

lepistone committed:
warning if we have a portfolio without agreement

lepistone committed:
open relevant POs from portfolio

lepistone committed:
shipment origin -> origin address

lepistone committed:
use only state to decide an agreement is suitable

We prefer doing a rough check and let the user make their choices
instead of risking to limit things too much for a small discrepancy.
Think of an agreement that would be good a few days later, or with a
slightly different incoterm.

lepistone committed:
add related origin address (shown in view)

lepistone committed:
doc for agreement 3.0

lepistone commented:
@gurneyalex I removed the WIP label and updated the README.

yvaucher commented on framework_agreement/model/portfolio.py:
```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+        self.ensure_one()
+        AGDates = namedtuple('AGDates', ['now', 'start', 'end'])
+        now = fields.date.today()
+        start = datetime.strptime(self.start_date, DATE_FORMAT).date()
```
Please use `fields.Datetime.from_string()`

yvaucher commented on framework_agreement/model/portfolio.py:
```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+    @api.multi
+    def is_valid_at_date(self, proposed_date):
+        self.ensure_one()
+        start = datetime.strptime(self.start_date, DATE_FORMAT)
```
same here for DATE_FORMAT

yvaucher commented on framework_agreement/model/portfolio.py:
```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+        'portfolio_id',
+        'portfolio_id.purchase_ids',
+        'portfolio_id.purchase_ids.portfolio_id',
+        'portfolio_id.purchase_ids.state',
```
`portfolio_id.purchase_ids.portfolio_id` aleady implies
`portfolio_id.purchase_ids` and `portfolio_id`

nbessi replied:
@yvaucher did this work on a one2many? if yes thats cool

yvaucher commented on framework_agreement/model/pricelist.py:
```diff
@@ -16,75 +14,419 @@
-from openerp.osv import orm, fields
+import time
+from openerp import models, fields, api, tools, exceptions, osv
+from openerp.tools.translate import _
```
You can add `_` in import from `openerp`

yvaucher commented on framework_agreement/model/pricelist.py:
```diff
@@ -16,75 +14,419 @@
-    """Add framework agreement behavior on pricelist"""
+    portfolio_id = fields.Many2one(
+        'framework.agreement.portfolio',
+        'Portfolio',
```
Maybe a help tooltip would be usefull to remind this field set means pricelist is an agreement.

yvaucher commented on framework_agreement/model/pricelist.py:
```diff
@@ -16,75 +14,419 @@
+                              products_by_qty_by_partner, context=None):
+        """Ugly duplication to implement the boolean use_agreement_prices."""
+        context = context or {}
+        date = context.get('date') or time.strftime('%Y-%m-%d')
```
Here as using old API you can use `DEFAULT_SERVER_DATE_FORMAT`

yvaucher commented on framework_agreement/model/pricelist.py:
```diff
@@ -16,75 +14,419 @@
+        return results
+
+    @api.model
+    def XXX_price_rule_get_multi(self, pricelist, products_by_qty_by_partner):
```
XXX is this a WIP ?

nbessi replied:
@yvaucher Work In Progess ? It looks more like forgotten code with call to super

yvaucher commented on framework_agreement/model/purchase.py:
```diff
@@ -84,120 +78,38 @@ def onchange_partner_id(self, partner_id):
-        'Agreement',
-        domain=[('portfolio_id', '=', 'order_id.portfolio_id')],
-    )
+    def insufficient_agreed_quantity(self):
```
Missing `@api.multi`

yvaucher commented on purchase_exceptions/model/purchase.py:
```diff
@@ -0,0 +1,219 @@
+        if default is None:
+            default = {}
+        default.update({
+            'ignore_exceptions': False,
```
This should go as `copy` attribute of `ignore_exceptions`

yvaucher commented:
Just some nitpicking to do. Otherwise LGTM

nbessi commented on framework_agreement/model/portfolio.py:
```diff
@@ -13,8 +13,15 @@
 
-from openerp import models, fields, api
+from openerp import models, fields, api, exceptions
+from openerp.tools import DEFAULT_SERVER_DATE_FORMAT as DATE_FORMAT
```
I guess this should not be used any more if the fields.Date helper are used

nbessi commented on framework_agreement/model/portfolio.py:
```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
                                  default=_company_get)
-    agreement_ids = fields.One2many('framework.agreement', 'portfolio_id',
-                                    'Agreements')
+    pricelist_ids = fields.One2many('product.pricelist', 'portfolio_id',
```
Should we set pricelist as mandatory?

nbessi commented on framework_agreement/model/portfolio.py:
```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+
+        Boiler plate as OpenERP returns string instead of date/time objects...
+
+        :returns: namedtuple('AGDates', ['now', 'start', 'end'])
```
I see some of my code here maybe we should alter the author ?

nbessi commented on framework_agreement/model/portfolio.py:
```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
 
     _sql_constraints = [
         ('uniq_portfolio',
          'unique(supplier_id, company_id)',
```
What if a portfolio is expired? You will not be able to create a new one? You should add a more complex check that uses the state.

nbessi commented on framework_agreement/model/portfolio.py:
```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+    def get_line_for_product(self, product, quantity=0):
+        self.ensure_one()
+        for product_line in self.line_ids:
+            if (product_line.product_id == product and
```
Hmm I guess we have an issue here. Nothing prevent to have many lines with same product in a portfolio so the chosen algo seems a bad fit or a constraint is missing.

nbessi commented on framework_agreement/model/portfolio.py:
```diff
@@ -35,15 +42,212 @@ def get_from_supplier(self, supplier):
+            'company': self.portfolio_id._company_get().id,
+            'product': self.product_id.id,
+        })
+        used_amount = self.env.cr.fetchone()[0] or 0
```
You do not take in account the use case where you SQL do not have a match and is None

nbessi commented on framework_agreement/model/pricelist.py:
```diff
@@ -16,75 +14,419 @@
+        'Incoterm',
+        help="International Commercial Terms are a series of predefined "
+        "commercial terms used in international transactions.")
+    origin_address_id = fields.Many2one('res.partner', 'Origin Address')
```
I do not think this address/portfolio data are really related to framework agreement/portfolio process. I guess I would spit this in a separated module. Agreement is mostly related to price negotiation. Most of the following is related to delivery management

yvaucher replied:
If this changes it is needed to adapt https://github.com/OCA/vertical-ngo/pull/187

nbessi commented on framework_agreement/model/pricelist.py:
```diff
@@ -16,75 +14,419 @@
+
+        version = False
+        for v in pricelist.version_id:
+            if (
```
I will add a function on pricelist.version model. That typically the kind of test you want to override. It will also be more readable

nbessi commented on framework_agreement/model/pricelist.py:
```diff
@@ -16,75 +14,419 @@
-                )
-                res[pricelist_id] = agreement.get_price(qty, currency=currency)
-        return res
+        products = map(lambda x: x[0], products_by_qty_by_partner)
```
Ok I see you where not able to use super and have to copy the whole function. But in this case you should clearly mark the code you have modifed.

Also I will in any case do a call to super if the boolean is not available. It minimise side effect in case of addons chain override. If it is really not possible you should do a monkey patch. This way your override will be at the bottom of stack.

Ok it is bad if two addons does the same, but I can live with that

nbessi commented on framework_agreement/model/pricelist.py:
```diff
@@ -16,75 +14,419 @@
+                                      related='suppinfo_id.product_tmpl_id',
+                                      readonly=True,
+                                      store=True)
+    pricelist_id = fields.Many2one(
```
why store every related firelds?

yvaucher replied:
For search I guess

nbessi commented on framework_agreement/model/purchase.py:
```diff
@@ -84,120 +78,38 @@ def onchange_partner_id(self, partner_id):
-        readonly=True,
-        related='order_id.portfolio_id',
-    )
+        if context.get('portfolio_id') and not context.get('pricelist_id'):
```
Can you explain this line ?

nbessi replied:
Ok I saw the new view

nbessi commented on purchase_exceptions/wizard/purchase_exception_confirm.py:
```diff
@@ -0,0 +1,48 @@
+    ignore = fields.Boolean('Ignore Exceptions')
+
+    @api.model
+    def default_get(self, field_list):
```
Why don't you use the default function of the fields?

nbessi commented:
Hello,

Thanks for the huge work. It is a radicval change in term of desing but it is clear it can cover more use cases even if the data input will probably become more tedious. 

It is a huge change in structure for an addons in. I wonder if it should not be a new addon ? Also there is no migration steps provided in the module that not really nice and it can totally break a installed version.

So without migration script or a way to warn the end user I have some hesitation to merge this PR.

I found some issues here and there that should be addresses. Please see in line comments

nbessi commented:
@lepistone any news ?

lepistone commented:
@nbessi, I saw your comments, thanks for your time! I'll definitely go through them at some point.

lepistone committed:
choose only portfolio with given product

Partially reverts 7536f28ecc1d.

lepistone committed:
add multi decorator

That way old-api code from the framework_agreement_sourcing module can
call this method.

StefanRijnhart commented:
What is the future of this PR? A lot of comments have remained unaddressed. I am currently improving some subtle usability issues with the frameworks module, but that will lead to conflicts here.

As for making this an additional module, that does not seem good to me. The original framework module will be obsolete anyways because the module has no sponsor anymore. I would much rather see this PR pick up the migration from the previous version.

gurneyalex commented:
@StefanRijnhart I'd like to see this merged. We are deploying it for a customer, but the project is on pause until september so it's hard to find the time to fix this with the other customers pression. I'll try to make some space for @lepistone to brush it up. Reviews are welcome.

lepistone commented:
Hi @StefanRijnhart @nbessi @yvaucher

I'm sorry I didn't follow up on this PR, yet.

I'm happy (and I think I can find the time) to address the most of the issues here. On the other hand, I am skeptical about being able to provide a comprehensive migration script. This PR is basically a breaking redesign, for a similar use case, but with different architecture and usability.
We don't have a need for migrating this module internally, so I'm unsure how to address that.

I agree with @StefanRijnhart that the old version this PR replaces is worse, and it is not maintained, at least not by us (since we are using the one in this PR).

What do you suggest? Maybe I could add a migration that `Raises`?

StefanRijnhart commented:
Thanks for the info. So the 7.0 version is also deprecated then, as far as you are concerned?

Are you sure that the migration is going to be such a big problem? What would break a mapping from framework agreements to agreement lines and pricelists?

lepistone commented:
@StefanRijnhart I gave a look to the data model before and after this PR, and I'm still pessimistic. The mapping might even be not well defined. In v2 we had to choose an agreement in every POL, because an agreement was specific to a product. In v3 we don't: we just choose a pricelist in the PO and that's it.

To migrate the configuration we'd need to create a portfolio for every old portfolio, then an agreement.product.line for every old agreement (but many fields don't map). Then for the prices itself it is even more complicated, because before we had agreement.lines and agreement.pricelists, and now we have product.supplierinfo (with a new relationship to pricelists), and partnerinfo containing the prices. Moreover the special pricelists need to be properly configured with the new rule added in this PR.

StefanRijnhart commented on framework_agreement/model/purchase.py:
```diff
@@ -26,47 +26,42 @@ class PurchaseOrder(models.Model):
+            # self.write does not work in an onchange
+            field_value = agreement[field_name]
+            if field_value:
+                self[field_name] = field_value
```
Getting KeyError: 'origin_address_id' at this point if I select a pricelist related to a portfolio, because this field does not exist on the purchase order.

lepistone replied:
You're right, I was missing a dependency. I added it, let's see if it builds.

StefanRijnhart replied:
I see you added a dependency on purchase_transport_multi_address. This strikes me a something that had better be put in a glue module, instead of including all of this in the basic framework module. My customer is interested in framework agreements without these extra flows. Would you consider such a refactoring?

lepistone replied:
Sure, nothing against that.

StefanRijnhart replied:
Here you go then https://github.com/lepistone/purchase-workflow/pull/1

lepistone committed:
bump version: this is agreement v3

lepistone committed:
depend on purchase_transport_multi_address

This is needed to handle origin_address_id in framework agreements v3.

StefanRijnhart commented on framework_agreement/view/purchase_view.xml:
```diff
@@ -21,29 +26,15 @@
-        <field name="price_unit" position="attributes">
-          <attribute name="on_change">1</attribute>
+        <field name="incoterm_id" position="attributes">
+          <attribute name="attrs">{'readonly': [('portfolio_id', '!=', False)]}</attribute>
```
When selecting a pricelist from a portfolio with an incoterm_id, the incoterm_id is visible after the onchange but as a readonly field it is not stored.

Commit:
[RFR] Glue module for framework_agreement + purchase_transport_multi_address

Commit:
[IMP] Don't overwrite any preexisting values
[IMP] Description

Commit:
[UPD] Summary

lepistone committed:
Merge pull request #1 from StefanRijnhart/8.0-framework_agreement_transport_multi_address

[RFR] Glue module for framework_agreement + purchase_transport_multi_address

StefanRijnhart commented:
@lepistone can we talk data model? We are actually less happy with the changes in the data model here. We have several scenarios that cannot be dealt with anymore:
- Have agreements for several products with the same supplier over different (overlapping) timespans while being able to order these products together on the same purchase order.
- Mix agreement products and regular products (prices for regular products are no longer retrieved. Can probably be fixed by reconfiguring the associated pricelists).

The old version of the module, with its configuration per purchase order line seems to support these scenarios fine. Was the simplification supposed to fix a usability issue or were there fundamental problems with the pricing mechanism in the old module?

lepistone commented:
I'll ping @jgrandguillaume (thanks Joël!)

StefanRijnhart commented:
Hi @lepistone, @jgrandguillaume any idea yet about the correct behaviour/policy of this module should be with respect to:
- Mixing multiple blanket orders and products without a blanket order within the same purchase order
- Have agreements for several products with the same supplier across different (overlapping) timespans

jgrandguillaume commented:
Hi @StefanRijnhart @lepistone 

I'm happy to have some interest on this topic ! First let me share with you the googledoc here:

https://docs.google.com/document/d/17Lm3axityll1oQ8KVxnm3RyJFyoKs3xNA2WXeKpgwmg/edit#

The main issue we faced with old design of FA was it was not based on the price-list mechanism. Using it was very difficult and some use cases wasn't supported (like multi-currency support for a given product and given supplier). Now we can handle that easily with one portfolio, various pricelist (FA) per currency / country, etc...( common denominator). Moreover, the previous design was a huge mess in term of UI and on_change management + very difficult to maintain (changing price, qty etc.. was a pain).

On your concerns Stefan:
- Mixing multiple blanket orders and products without a blanket order within the same purchase order

=> Not sure I get your point here. But if you speak of having two line in a same PO for the same product, one under agreement, the other not, then you're right, the price computation won't work. But you can still change the price manually (which wasn't possible with first design). All in all, it is good practice to have one PO for good under FA and another for goods not in the FA.

Also, note that as the FA are now pricelist, you can define price for product in that pricelist that are not part of the agreement. e.g.

 Portfolio
    \* Product A ; 2000
    \* Product B ; 3000
 FA 1 Price for China:
    \* Product A 10.-
    \* Product B 20.-
    \* Product C 50.- (not in agreement, but still a price for this supplier)
 FA 2 price for USA:
- Product A 15.-
- Product B 30.-
- Product C 80.- (not in agreement, but still a price for this supplier)
- Have agreements for several products with the same supplier across different (overlapping) timestamps

=> I think you can do this. You can have an agreement portfolio based on various product. You can define various portfolio for different time period for the very same supplier => done !

I hope this helps. I'm really interested to have this PR merged even if no migration script as no-one seems to maintain version 7.0 any-more except us ;)

Regards,

Joël

StefanRijnhart commented:
Thanks Joël for your extensive clarifications!

If I remember correctly, products outside the portfolio will return a zero price. Maybe each portfolio pricelist version should be created with a low priority item referring to the default purchase pricelist. I'll check this.

Having multiple portfolios for the same supplier (to allow for overlapping agreements for different products) is currently explicitely blocked by a constraint. If you agree (@lepistone?), this constraint can simply be removed.

Having overlapping agreements for different products in the same portfolio would be ideal for us, but this does not fit the Odoo 8.0 data model of pricelists and versions. I guess we could repartition a portfolio's pricelist versions according to the dates of the underlying agreements on every update, but this would be elaborate if not artificial and ugly. On our side, I think we would rather wait with improving the module on this point until we move to version 9 where the dates are on the pricelist items, which seems to suit this business case perfectly. Until then, I will through some purchase exceptions at the users if products are detected from another portfolio.

@lepistone, for easier processing by the wider community, should we split off the purchase_exceptions bit and propose that separately?

lepistone commented:
Hi @StefanRijnhart! I'm sorry this was abandoned here. The project for which we developed this is finished, so as mentioned before I don't have time and budget to do much here. Are you still working on it, you made a fork maybe?

If there is interest here I could just address some of the issues brought up in the reviews. Or if someone wants to adopt this and resubmit it to the OCA, of course you're welcome. Thanks!

StefanRijnhart commented:
Hi @lepistone, thanks for the update. As this solution was overkill for us with additional problems, we went for a much simpler customization instead. This should probably be closed then.
